### PR TITLE
Adds forgotPassword/confirmForgotPassword

### PR DESCRIPTION
**Pull request by TrekSoft**
https://quip-amazon.com/V550Arr1aKJN/Amplify-Auth-Category-Interface#KVW9CAtlT6x

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.java [hunk 1]
```diff
@@ -367,22 +370,88 @@ public void onResult(UserStateDetails result) {
...
+            @NonNull Consumer<AuthException> onException
+    ) {
+        AWSMobileClient.getInstance().confirmForgotPassword(
+            newPassword,
+            confirmationCode,
+            new Callback<ForgotPasswordResult>() {
+                @Override
+                public void onResult(ForgotPasswordResult result) {
+                    if (result.getState().equals(ForgotPasswordState.DONE)) {
+                        onSuccess.call();
```
Comment (jamesonwilliams) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.java:433]: How come you don't pass back `convertCodeDeliveryDetails(result.getParameters())` here too? Doesn't contain any valuable info?

File: core/src/main/java/com/amplifyframework/auth/AuthCategoryBehavior.java [hunk 2]
```diff
@@ -33,7 +34,7 @@
      * Creates a new user account with the specified username and password.
      * Can also pass in user attributes to associate with the user through
      * the options object.
-     * @param username This can be a typical username, email, phone number, etc.
+     * @param username This can be a typical username, email, phone number, etc. depending on the configuration
```
Comment (jamesonwilliams) [inline core/src/main/java/com/amplifyframework/auth/AuthCategoryBehavior.java:37]: For both places, maybe:
```
@param username A login identifier e.g. `awsuser22`; or, could be an email or phone number, depending on configuration
```

File: core/src/main/java/com/amplifyframework/auth/AuthCategoryBehavior.java [hunk 3]
```diff
@@ -154,4 +155,28 @@ void signInWithFacebook(
...
+     * @param onSuccess Success callback
+     * @param onError Error callback
+     */
+    void forgotPassword(
+            @NonNull String username,
+            @NonNull Consumer<AuthCodeDeliveryDetails> onSuccess,
+            @NonNull Consumer<AuthException> onError);
+
+    /**
+     * Complete password recovery process by inputting user's desired new password and confirmation code.
```
Comment (jamesonwilliams) [inline core/src/main/java/com/amplifyframework/auth/AuthCategoryBehavior.java:171]: `and the confirmation code that came from (where does it come from)`?

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.java (hunk 1 above)
Comment (TrekSoft) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.java:433]: Yeah no it's a messed up implementation in AWSMobileClient where there's only one legitimate value which is Done and everything else is empty or an error. :P

File: core/src/main/java/com/amplifyframework/auth/AuthCategoryBehavior.java (hunk 2 above)
Comment (TrekSoft) [reply core/src/main/java/com/amplifyframework/auth/AuthCategoryBehavior.java:37]: Sure, will add this in the next round

File: core/src/main/java/com/amplifyframework/auth/AuthCategoryBehavior.java (hunk 3 above)
Comment (TrekSoft) [reply core/src/main/java/com/amplifyframework/auth/AuthCategoryBehavior.java:171]: It depends how it's configured so I think I'll leave it as is.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.java (hunk 1 above)
Comment (jamesonwilliams) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.java:433]: Haha oh alright!

**Pull request merged**